hphp/runtime/vm/jit/vasm-arm.cpp

```diff
@@ -1487,77 +1483,71 @@ void lower(const VLS& e, pushpm& i, Vlabel b, size_t z) {
-    v << copy{i.s, d};
-    v << movtqb{d, i.d};
+    v << copy{i.s, i.d};
   });
 }
 
-void lower(const VLS& e, movtdq& i, Vlabel b, size_t z) {
-  lower_impl(e.unit, b, z, [&] (Vout& v) {
-    v << copy{i.s, i.d};
-  });
+void lower(const VLS& e, movzbw& i, Vlabel b, size_t z) {
+  lower_movz(e, i, b, z);
 }
 
-void lower(const VLS& e, cmpb& i, Vlabel b, size_t z) {
-  lower_impl(e.unit, b, z, [&] (Vout& v) {
-    auto s0 = v.makeReg();
-    auto s1 = v.makeReg();
-    v << movzbl{i.s0, s0};
-    v << movzbl{i.s1, s1};
-    v << cmpl{s0, s1, i.sf};
-  });
+void lower(const VLS& e, movzbl& i, Vlabel b, size_t z) {
+  lower_movz(e, i, b, z);
 }
 
-void lower(const VLS& e, cmpbi& i, Vlabel b, size_t z) {
-  lower_impl(e.unit, b, z, [&] (Vout& v) {
-    auto s1 = v.makeReg();
-    v << movzbl{i.s1, s1};
-    v << cmpli{i.s0, s1, i.sf};
-  });
+void lower(const VLS& e, movzbq& i, Vlabel b, size_t z) {
+  lower_movz(e, i, b, z);
 }
 
-#define Y(vasm_opc, conv_opc, load_opc, cmp_opc)            \
-void lower(const VLS& e, vasm_opc& i, Vlabel b, size_t z) { \
-  lower_impl(e.unit, b, z, [&] (Vout& v) {                  \
-    auto s0 = v.makeReg();                                  \
-    v << conv_opc{i.s0, s0};                                \
-    lowerVptr(i.s1, v);                                     \
-    auto s1 = v.makeReg();                                  \
-    v << load_opc{i.s1, s1};                                \
-    v << cmp_opc{s0, s1, i.sf};                             \
-  });                                                       \
+void lower(const VLS& e, movzwl& i, Vlabel b, size_t z) {
+  lower_movz(e, i, b, z);
 }
 
-Y(cmpbm, movzbl, loadzbl, cmpl)
-Y(cmpwm, movzwl, loadw, cmpl)
+void lower(const VLS& e, movzwq& i, Vlabel b, size_t z) {
+  lower_movz(e, i, b, z);
+}
 
-#undef Y
+void lower(const VLS& e, movzlq& i, Vlabel b, size_t z) {
+  lower_movz(e, i, b, z);
+}
 
-void lower(const VLS& e, testb& i, Vlabel b, size_t z) {
+void lower(const VLS& e, movtdb& i, Vlabel b, size_t z) {
   lower_impl(e.unit, b, z, [&] (Vout& v) {
-    if (i.s0 == i.s1) {
-      v << testbi{(uint8_t)0xff, i.s1, i.sf};
-    } else {
-      auto s0 = v.makeReg();
-      auto s1 = v.makeReg();
-      v << movzbl{i.s0, s0};
-      v << movzbl{i.s1, s1};
-      v << testl{s0, s1, i.sf};
-    }
+    auto d = v.makeReg();
+    v << copy{i.s, d};
+    v << movtqb{d, i.d};
   });
 }
 
-void lower(const VLS& e, testbi& i, Vlabel b, size_t z) {
+void lower(const VLS& e, movtdq& i, Vlabel b, size_t z) {
   lower_impl(e.unit, b, z, [&] (Vout& v) {
-    auto s1 = v.makeReg();
-    v << movzbl{i.s1, s1};
-    v << testli{i.s0, s1, i.sf};
+    v << copy{i.s, i.d};
   });
 }
 
+template<typename load_op, typename cmp_op, typename cmpm>
+void lower_cmpm(const VLS&e, cmpm& i, Vlabel b, size_t z) {
+  lower_impl(e.unit, b, z, [&] (Vout& v) {
+    lowerVptr(i.s1, v);
+    auto tmp0 = v.makeReg();
+    auto tmp1 = v.makeReg();
+    v << copy{i.s0, tmp0};
```

Comment:
Because `cmpbm.s0` is a Vreg8 and `cmpl.s0` is a Vreg32 and zero-extension in the lowering is replaced by `copy{}`.
Same is true for `cmpwm.s0` (Vreg16). I.e. it does not compile without the `copy{}` or a zero-extension. And since `lowerForARM()` is called before `optimizeCopies()` the `copy{}` will be eliminated anyways.